zilliqa/src/node.rs

```diff
@@ -399,4 +442,68 @@ impl Node {
 
         Ok(())
     }
+
+    fn handle_blocks_request(&mut self, source: PeerId, request: BlocksRequest) -> Result<()> {
+        let block = match request.0 {
+            crate::message::BlockRef::Hash(hash) => self.consensus.get_block(&hash),
+            crate::message::BlockRef::View(view) => self.consensus.get_block_by_view(view),
+            crate::message::BlockRef::Number(number) => self.consensus.get_block_by_number(number),
+        }?;
+
+        let block = match block {
+            Some(block) => block,
+            None => {
+                debug!("ignoring blocks request for unknown block: {:?}", request.0);
+                return Ok(());
+            }
+        };
+
+        let block_number = block.header.number;
+        let mut blocks: Vec<Block> = Vec::new();
+
+        for i in block_number..block_number + 100 {
+            let block = self.consensus.get_block_by_number(i);
+            if let Ok(Some(block)) = block {
+                blocks.push(block);
+            } else {
+                break;
+            }
+        }
+
+        trace!(
+            "Responding to new blocks request of {:?} starting {} with {} blocks",
+            request,
+            block_number,
+            blocks.len()
+        );
+
+        self.message_sender.send_external_message(
+            source,
+            ExternalMessage::BlocksResponse(BlocksResponse { blocks }),
+        )?;
+
+        Ok(())
+    }
+
+    fn handle_blocks_response(&mut self, _: PeerId, response: BlocksResponse) -> Result<()> {
+        trace!(
+            "Received blocks response of length {}",
+            response.blocks.len()
+        );
+        let mut was_new = false;
+
+        for block in response.blocks {
+            was_new = self.consensus.receive_block(block)?;
+        }
+
+        if was_new {
```

Comment:
effectively, every time you verify that you got new blocks when syncing from a peer, ask for more with your new head as the tip. This way it recurses until you should have them all.

Reply:
Nice!